datumaro/plugins/icdar_format/extractor.py

```diff
@@ -0,0 +1,173 @@
+# Copyright (C) 2020 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import os.path as osp
+from glob import glob
+
+from datumaro.components.extractor import (AnnotationType, Bbox, DatasetItem,
+    Importer, Label, LabelCategories, Points, SourceExtractor)
+from datumaro.plugins.icdar_format.format import IcdarPath, IcdarTask
+
+
+class _IcdarExtractor(SourceExtractor):
+    def __init__(self, path, task):
+        if task is IcdarTask.word_recognition:
+            if not osp.isfile(path):
+                raise Exception("Can't read annotation file '%s'" % path)
+            subset = osp.basename(osp.dirname(path))
+            self._dataset_dir = osp.dirname(osp.dirname(osp.dirname(path)))
+        elif task is IcdarTask.text_localization:
+            if not osp.isdir(path):
+                raise Exception("Can't open folder with annotation files'%s'" % path)
+            subset = osp.basename(path)
+            self._dataset_dir = osp.dirname(osp.dirname(path))
+        self._path = path
+        self._task = task
+        super().__init__(subset=subset)
+
+        self._categories = self._load_categories()
+        self._items = list(self._load_items().values())
+
+    def _load_categories(self):
+        categories = {}
+        path = osp.join(self._dataset_dir, IcdarPath.TASK_DIR[self._task],
+            IcdarPath.VOCABULARY_FILE)
+        labels = []
+        if osp.isfile(path):
+            with open(path, encoding='utf-8') as labels_file:
+                labels = [s.strip() for s in labels_file]
+        else:
+            if self._task is IcdarTask.word_recognition:
+                paths = [self._path]
+            else:
+                paths = [p for p in glob(osp.join(self._path, '*.txt'),
+                    recursive=True)]
+            for path in paths:
+                with open(path, encoding='utf-8') as f:
+                    for line in f:
+                        objects = line.split()
+                        if len(objects) == 1:
+                            objects = line.split(',')
+
+                        if len(objects) == 9:
+                            label = objects[8]
+                        elif len(objects) == 5:
+                            label = objects[4]
+                        elif len(objects) == 2:
+                            label = objects[1]
+                        else:
+                            continue
+                        if label[0] == '"' and label[-1] == '"':
+                            label = label[1:-1]
+                        labels.append(label)
+
+        categories[AnnotationType.label] = \
+            LabelCategories().from_iterable(sorted(labels))
+
+        return categories
+
+    def _load_items(self):
```

Comment:
It seems like there is nothing in common in these subformats. Wouldn't it be better to split their implementations?